src/logits_processor.cpp

```diff
@@ -0,0 +1,235 @@
+      // when logits processor decides to stop, we mask all tokens except the EOS token
+      mask = std::vector<uint32_t>((params_->config.model.vocab_size - 1) / 32 + 1, 0);
+      uint32_t eos_mask32 = 1 << (eos_token_ % 32);
+      mask[eos_token_ / 32] = eos_mask32;
+    } else {
+      mask.reserve((params_->config.model.vocab_size - 1) / 32 + 1);
+      for (int i = 0; i < (params_->config.model.vocab_size - 1) / 32 + 1; i++) {
+        mask.push_back(mask_result.sample_mask[i]);
+      }
+    }
+    masks.push_back(mask);
+  }
+  return masks;
+}
+
+void GuidanceLogitsProcessor::CommitTokens(std::span<int32_t> tokens) {
+  for (int i = 0; i < params_->search.batch_size; i++) {
+    LlgCommitResult commit_result;
+    auto error = llg_commit_token(llg_constraints_[i].get(), static_cast<uint32_t>(tokens[i]), &commit_result);
+    if (error != 0) {
+      std::string error_message = llg_get_error(llg_constraints_[i].get());
+      throw std::runtime_error("Error committing tokens: " + error_message);
+    }
+  }
+  mask_future_ = std::async(std::launch::async, [&]() {
+    return ComputeMask();
+  });
+  masks_.clear();
+}
+
+std::vector<std::vector<uint32_t>> GuidanceLogitsProcessor::GetMask() {
+  if (masks_.empty()) {
+    masks_ = mask_future_.get();
+  }
+  return masks_;
+}
+
+void GuidanceLogitsProcessor::ProcessLogits(DeviceSpan<float> logits) {
+  auto masks = GetMask();
+
+  if (params_->p_device->GetType() == DeviceType::CUDA) {
+    const size_t words_per_row = params_->config.model.vocab_size / 32;
+    const size_t total_words = masks.size() * words_per_row;
+    std::vector<uint32_t> flat_masks(total_words);
+    uint32_t* dst = flat_masks.data();
+    for (const auto& row : masks) {
+      std::memcpy(dst, row.data(), words_per_row * sizeof(uint32_t));
+      dst += words_per_row;
+    }
+    auto cuda_logits_mask_ptr_ = params_->p_device->Allocate<uint32_t>(total_words);
+    copy(std::span<const uint32_t>{flat_masks}, cuda_logits_mask_ptr_.CpuSpan());
+    cuda_logits_mask_ptr_.CopyCpuToDevice();
+    params_->p_device->LaunchAddLogitsMask(logits.Span().data(), params_->search.batch_size, params_->config.model.vocab_size, cuda_logits_mask_ptr_.Span().data());
+    return;
+  }
+  size_t vocab_index = 0;
+
+  auto logits_span = logits.CpuSpan();
+  for (int index = 0; index < params_->search.batch_size; index++) {
+    auto subspan = logits_span.subspan(vocab_index, params_->config.model.vocab_size);
+    auto& mask = masks[index];
+    for (size_t i = 0; i < params_->config.model.vocab_size; i++) {
+      // mask is a 32-bit integer, where each bit corresponds to a token in the vocabulary.
+      // If the bit is set, the corresponding token is masked (i.e., its logit is set to the lowest possible value).
+      subspan[i] = mask[i / 32] & (1 << (i % 32)) ? subspan[i] : std::numeric_limits<float>::lowest();
+    }
+    vocab_index += params_->config.model.vocab_size;
+  }
+}
+
+void GuidanceLogitsProcessor::Reset() {
```

Comment:
Can there be a comment saying what doesn't get reset to the same state as a freshly constructed GuidanceLogitsProcessor?